File: GFG POTD/019-June-2025/003-count-number-of-substrings.cpp
// https://www.geeksforgeeks.org/problems/count-number-of-substrings4528/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    int getCntSubstr(string &s, int n, int k) {
        int cntSubstr = 0;
        int i = 0, j = 0;
        vector<int> freq(26, 0);
        int count = 0;
        while (j < n) {
            freq[s[j] - 'a'] ++;
            if (freq[s[j] - 'a'] == 1) count ++;
            while (i <= j && count > k) {
                freq[s[i] - 'a'] --;
                if (!freq[s[i] - 'a']) count --;
                i ++;
            }
            cntSubstr += (j - i + 1);
            j ++;
        }
        return cntSubstr;
    }
    
    // TC : O(N)
    // SC : O(1)
    int approach_1(string &s, int k) {
        int n = s.length();
        int totalCntK = getCntSubstr(s, n, k);  // cnt of substrings with at most k-distinct characters
        int remainCnt = getCntSubstr(s, n, k - 1); // cnt of substrings with at most (k - 1)distinct characters
        
        return (totalCntK - remainCnt); // cnt of exactly k substrings
    }
    
    // TC : O(N)
    // SC : O(1)
    int approach_2(string &s, int k) {
        int n = s.length();
        int i = 0, j = 0, prev_i = 0;
        int cntSubStr = 0;
        unordered_map<char, int> mpp;
        int count = 0;
        
        while (j < n) {
            mpp[s[j]] ++;
            
            while (mpp.size() > k) {
                mpp[s[i]] --;
                if (!mpp[s[i]]) mpp.erase(s[i]);
                i ++;
                prev_i = i; 
            }
            
            while (mpp[s[i]] > 1) { // shrink if char pointed by i has freq > 1
                mpp[s[i]] --;
                i ++;
            }
            
            if (mpp.size() == k) 
                cntSubStr += (i - prev_i + 1);
            
            j ++;
        }
        
        return cntSubStr;
    }
  public:
    int countSubstr(string& s, int k) {
        // return approach_1(s, k);
        return approach_2(s, k);     // ONE PASS
    }
};

File: GFG POTD/019-June-2025/003-count-number-of-substrings.java
// https://www.geeksforgeeks.org/problems/count-number-of-substrings4528/1




import java.util.*;

// User function Template for Java

class Solution {
    int countSubstr(String s, int k) {
        // return approach_1(s, k); // 2-pass soln.
        return approach_2(s, k); // 1-pass soln.
    }
    
    // TC : O(N)
    // SC : O(1) 
    int approach_2(String s, int k) {
        int n = s.length();
        List<Integer> freq = new ArrayList<>(Collections.nCopies(26, 0));
        int cntFreq = 0;
        
        int i = 0, j = 0, prev_i = 0;   // prev_i will track index of starting valid substring
        int cntSubstr = 0;
        
        while (j < n) {
            int idx = s.charAt(j) - 'a';
            freq.set(idx, freq.get(idx) + 1);
            if (freq.get(idx) == 1) cntFreq += 1;
            
            while (cntFreq > k) {
                int chI = s.charAt(i) - 'a';
                freq.set(chI, freq.get(chI) - 1);
                if (freq.get(chI) == 0) {
                    cntFreq --;
                }
                i ++;
                prev_i = i;
            }
            
            while (i < j && freq.get(s.charAt(i) - 'a') > 1) {
                int chI = s.charAt(i) - 'a';
                freq.set(chI, freq.get(chI) - 1);
                i ++;
            }
            
            if (cntFreq == k) cntSubstr += (i - prev_i + 1);
            j ++;
        }
        
        return cntSubstr;
    }
    
    // TC : O(N)
    // SC : O(1)
    int approach_1(String s, int k) {
        int totalCnt = getSubStringCnt(s, k);       // substrings contains distinct chars at most k-times
        int remains  = getSubStringCnt(s, k - 1);   // substrings contains distinct chars at most k-1 times
    
        return totalCnt - remains;  // substrings contains distinct chars exactly k-times
    }
    
    // TC : O(N)
    // SC : O(1)
    int getSubStringCnt(String s, int k) {
        int n = s.length();
        
        int[] freq = new int[26];
        for (int i = 0; i < 26; i ++) {
            freq[i] = 0;
        }
        
        int i = 0, j = 0, ans = 0;
        int cntFreq = 0;
        
        while (j < n) {
            freq[s.charAt(j) - 'a'] ++;
            if (freq[s.charAt(j) - 'a'] == 1) cntFreq += 1;
            
            while (i <= j && cntFreq > k) {
                freq[s.charAt(i) - 'a'] --;
                if (freq[s.charAt(i) - 'a'] == 0) {
                    cntFreq --;
                }
                i ++;
            }
            
            ans += (j - i + 1);
            j ++;
        }
        
        return ans;
    }
}